File: cs50x/week6/pset6/sentimental-credit/credit.py
# TODO
import sys


def main():
    try:
        credit_card = int(input("Number: "))
    except:
        print("Credit card not valid")
        sys.exit(1)
    # Check for Luhns Algorithmn
    if (not checksum(credit_card)):
        print("INVALID")
        sys.exit(0)
    # Check if AMEX
    if is_AMEX(credit_card):
        print("AMEX")
    # Check if Mastercard
    elif is_MC(credit_card):
        print("MASTERCARD")
    # Check if Visa
    elif is_VISA(credit_card):
        print("VISA")
    # Card is invalid
    else:
        print("INVALID")
    sys.exit(0)

# Length and Beginning of given credit card requirements


def is_AMEX(number):
    result = False
    number = str(number)
    length = len(number)
    if length == 15 and number[0] == "3" and (number[1] == "4" or number[1] == "7"):
        result = True
    return result

# Length and Beginning of given credit card requirements


def is_MC(number):
    result = False
    number = str(number)
    length = len(number)
    if length == 16 and (number[0:2] == "51" or number[0:2] == "52" or number[0:2] == "53" or number[0:2] == "54" or number[0:2] == "55"):
        result = True
    return result

# Length and Beginning of given credit card requirements


def is_VISA(number):
    result = False
    number = str(number)
    length = len(number)
    if (length == 16 or length == 13) and number[0] == "4":
        result = True
    return result


def checksum(number):
    """Luhn's Algorithmn """
    result = False
    if number < 0:
        return result
    length = len(str(number))
    sum2 = 0
    sum = 0

    for i in range(length):
        if i % 2 == 1:
            num = (number % 10)*2
            if num > 9:
                sum2 += num % 10
                num //= 10
            sum2 += num
        else:
            sum += number % 10
        number //= 10
    if (sum + sum2) % 10 == 0:
        result = True
    return result


if __name__ == '__main__':
    main()

File: cs50x/week1/pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int getLength(long number);
void getDigits(long number, int digits[]);
void identifyCard(int digitCount, int digits[]);
bool checkSum(long number, int digitCount);

int main(void)
{
    // Input Number
    long number = get_long("Number: ");
    // Number > 0, else Number is invalid
    if (number > 0)
    {
        // Get digitcount
        int digitCount = getLength(number);
        int digits[digitCount];
        // Get digits
        getDigits(number, digits);
        // Identification Credit Card
        if (checkSum(number, digitCount))
        {
            identifyCard(digitCount, digits);
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}
// Return the count of digits from a given number
int getLength(long number)
{
    int count = 0;
    while (number != 0)
    {
        number = number / 10;
        count++;
    }
    return count;
}
// Returns the pointer to the beginning of the array, cointaining digits of the given number
void getDigits(long number, int digits[])
{
    int i = 0;
    while (number != 0)
    {
        digits[i] = number % 10;
        number = number / 10;
        i++;
    }
}
//
void identifyCard(int digitCount, int digits[])
{
    // Amex
    if (digitCount == 15 && digits[digitCount - 1] == 3 && (digits[digitCount - 2] == 4 || digits[digitCount - 2] == 7))
    {
        printf("AMEX\n");
    }
    // VISA with 13 digits
    else if (digitCount == 13 && digits[digitCount - 1] == 4)
    {
        printf("VISA\n");
    }
    else if (digitCount == 16)
    {
        // VISA with 16 digits and constraints
        if (digits[digitCount - 1] == 4)
        {
            printf("VISA\n");
        }
        // MasterCard 16 digits
        else if (digits[digitCount - 1] == 5)
        {
            if (digits[digitCount - 2] == 1 || digits[digitCount - 2] == 2 || digits[digitCount - 3] == 1 || digits[digitCount - 2] == 4
                || digits[digitCount - 2] == 5)
            {
                printf("MASTERCARD\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}
bool checkSum(long number, int digitCount)
{
    bool result = false;
    // Array with saved digits
    int d[digitCount];
    int i = 0;
    while (number != 0)
    {
        d[i] = number % 10;
        number = number / 10;
        i++;
    }
    int sumP = 0;
    int sumS = 0;
    for (int k = 0 ; k < digitCount; k++)
    {
        // Double every 2nd digit and sum product digits
        if (k % 2 == 1)
        {
            int temp = d[k] * 2;
            while (temp != 0)
            {
                sumP += temp % 10;
                temp = temp / 10;
            }
        }
        else
        {
            sumS += d[k];
        }
    }
    int total = sumP + sumS;
    if (total % 10 == 0)
    {
        result = true;
    }
    return result;
}